src/liballoc/collections/btree/set.rs

```diff
@@ -1273,19 +1287,21 @@ impl<'a, T: Ord> Iterator for SymmetricDifference<'a, T> {
 
     fn next(&mut self) -> Option<&'a T> {
         loop {
-            match cmp_opt(self.a.peek(), self.b.peek(), Greater, Less) {
-                Less => return self.a.next(),
-                Equal => {
-                    self.a.next();
-                    self.b.next();
-                }
-                Greater => return self.b.next(),
+            match self.0.nexts() {
+                (None, None) => return None,
+                (Some(item), None) => return Some(item),
+                (None, Some(item)) => return Some(item),
```

Comment:
I like the @ thing. I don't think combining here helps readability because the repeated code is tiny and aligned, but I can live with it.

You raise an interesting name debate: I recently got to https://internals.rust-lang.org/t/the-is-not-empty-method-as-more-clearly-alternative-for-is-empty/10612/91 and took for granted the statement that we should be talking about items and not elements. Now I see elements in the comments too. But I wouldn't call an optional element an element, I'd call it `next`.

I'm going to benchmark too `if a_next.and(b_next).is_none() { return a_next.or(b_next); }`. Not a beauty but it is short.

Reply:
Option::xor is new. I'm mildly excited because it looks like it fits here

Reply:
Not sure how xor fits here? I used it instead of `or` here, but it doesn't serve any purpose there.

    let (a_next, b_next) = self.0.nexts();
    if a_next.and(b_next).is_none() {
        return a_next.xor(b_next)
    }

Now hold on to your seat (or be weary of benchmarks)... this xor thrashes performance by up to 20%.
Just turn that `xor` into `or` and performance is back to what it is with the 4 line match above.

Another disappointment:

   match self.0.nexts() {
        (None, None) => return None,
        (next @ Some(_), None) | (None, next @ Some(_)) => return next,
        (Some(_), Some(_)) => (),
    }
is about 6% worse, and more so with the "parted_" benches, where there are no (Some(_), Some(_)) matches. I simplified it further to (a surprise in itself):

   match self.0.nexts() {
        (Some(_), Some(_)) => (),
        (next, None) | (None, next) => return next,
    }
and that improves it to just about the best level, except for the "parted_" benches.

Reply:
Oh and earlier I had tried some alternatives to MergeIter::next returning a pair of items:

- (Option<&Item>, usize) where second is the number of sets the item (if any) was in: seemed okay
- (Option<&Item>, bool) where second is basically "the if clause of SymmetricDifference": ugly, even if I'd turn the bool into an enum
- move that bool into MergeIter, or just the next function itself: ugly

and none seemed to perform better or worse.

Reply:
I saw another angle to that xor: pass a lambda to the next function that says what to do with a result on both sides. Option::or for Union, Option::xor for SymmetridDifference. And also move the possible loop inside MergeIterInner. Looks nice, actually, but performance suffers (according to some benchmarking), also for Union.